### PR TITLE
[MRG+1] Nightly cron

### PR DESCRIPTION
**Pull request by aaronreidsmith**
# Description

This PR adds a nightly cron job to build `pmdarima` against all OS/Python versions we support using the **latest** version of all of our dependencies. It only builds wheel files, because if wheel files break, conda and source will break as well.

This PR adds the GitHub Actions workflow file, as well as a Python file used to find the latest versions of our dependencies.

I am planning on using [this action](https://github.com/marketplace/actions/slack-notify) to post to a slack channel if/when it fails, hence why this is a WIP (need to get all the slack secrets added)

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [X] The Python script successfully strips off pinned versions
- [X] The tests successfully run against latest versions

# Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by aaronreidsmith** [comment]
Per our Slack conversation, this has been updated to only run against latest everything (meaning latest OS, latest Python [currently 3.8], and latest dependencies)

File: .github/workflows/nightly_cron.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        os: [windows-latest, macos-latest, ubuntu-latest]
+        python-version: ['3.8']
+        architecture: ['x86', 'x64']
+        exclude:
+          # Don't build 32-bit on Mac or Linux
+          - os: macos-latest
+            architecture: x86
+
+          - os: ubuntu-latest
+            architecture: x86
```
Comment (tgsmith61591) [inline .github/workflows/nightly_cron.yml]: An aside, I really like the way GH actions structures their YAML. Very readable in plain English.

File: .github/workflows/nightly_cron.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      - name: Installing generated wheel
+        run: pip install --pre --no-index --find-links dist/ pmdarima
+        shell: bash
+
+      - name: Running unit tests
+        run: |
+          if [ "${{ matrix.os }}" == "macos-latest" ]; then
+            export PMD_MPL_BACKEND=TkAgg
+          fi
+          pytest --showlocals --durations=20 --pyargs pmdarima --benchmark-skip
```
Comment (tgsmith61591) [inline .github/workflows/nightly_cron.yml]: FWIW, you prob don't need the `--benchmark-skip` since we no longer use pytest-benchmark.

File: build_tools/github/get_latest_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    requirements = []
+    for requirements_file in requirements_files:
+        with open(requirements_file) as file:
+            for line in file:
+                requirement = line.strip()
+                if line.startswith('#'):
+                    continue
+                match = re.match(r'^([A-Za-z\-0-9]+)', requirement)
+                if match.group(0).lower() not in requirements:
+                    requirements.append(match.group(0).lower())
```
Comment (tgsmith61591) [inline build_tools/github/get_latest_dependencies.py]: I'm not super clear on this regex. But I get the purpose of the loop. FWIW (and it makes no real difference here because the size is so small and is not user-facing), set lookups are O(1) while list lookups are O(N)

Comment (aaronreidsmith) [reply build_tools/github/get_latest_dependencies.py]: I initially had a set, but our requirements file has a specific order that I want to maintain. It's important to install Cython and Numpy before anything else, and if I use a set and cast to a list, I can't guarantee that 😕

File: .github/workflows/nightly_cron.yml (hunk 2 above)
Comment (aaronreidsmith) [reply .github/workflows/nightly_cron.yml]: Cool, I will remove that!

File: .github/workflows/nightly_cron.yml (hunk 1 above)
Comment (aaronreidsmith) [reply .github/workflows/nightly_cron.yml]: Yeah, they make it pretty simple! Wish their docs were a little better though 🤔

File: build_tools/github/get_latest_dependencies.py (hunk 3 above)
Comment (aaronreidsmith) [reply build_tools/github/get_latest_dependencies.py]: Also, what this regex is doing is looking for any letter characters, dashes, or numbers starting at the beginning of the string. So, it will either match anything up until our pins (`>,<,~,=`), or until the end of the line

**Comment by aaronreidsmith** [comment]
Personally, I wouldn't put a badge on the readme, because I feel like we wouldn't be able to explain what it is actually representing in such a small space (the badge title). Also, since the distributions _will_ have pinned versions, it is kind of a moot point to display that. I think once I add the slack notifications, it should be good enough

**Comment by aaronreidsmith** [comment]
This should be good, now that I have added the notification step
